stream/replicator.h

```diff
@@ -403,15 +464,332 @@ struct StreamReplicatorImpl {
+    }
+  }
+
+  void FollowRemoteStream(const std::string& url, SubscriptionMode subscription_mode = SubscriptionMode::Unchecked) {
+    if (remote_follower_) {
+      CURRENT_THROW(StreamIsAlreadyFollowingException());
+    }
+
+    std::lock_guard<std::mutex> lock(mutex_);
+    const bool has_borrowed_publisher = Exists(borrowed_publisher_);
+    try {
+      remote_follower_ = std::make_unique<RemoteStreamFollower>(
+          has_borrowed_publisher ? std::move(Value(borrowed_publisher_)) : stream_->BecomeFollowingStream(),
+          url,
+          stream_->Data()->Size(),
+          stream_->Data()->CurrentHead() + std::chrono::microseconds(1),
+          subscription_mode);
+      borrowed_publisher_ = nullptr;
+    } catch (current::Exception& e) {
+      // Can't follow the remote stream for some reason,
+      // restore the stream state and propagate the exception.
+      if (has_borrowed_publisher)
+        borrowed_publisher_ = stream_->BecomeFollowingStream();
+      else
+        stream_->BecomeMasterStream();
+      throw;
+    }
+  }
+
+  void FlipToMaster(uint64_t key) {
+    if (stream_->IsMasterStream()) {
+      CURRENT_THROW(StreamIsAlreadyMasterException());
+    }
+    if (!remote_follower_) {
+      CURRENT_THROW(StreamDoesNotFollowAnyoneException());
+    }
+    remote_follower_->PerformMasterFlip(*Value(stream_), key);
+    remote_follower_ = nullptr;
+    stream_->BecomeMasterStream();
+  }
+
+  bool IsMasterStream() const { return stream_->IsMasterStream(); }
+  stream_t& Stream() { return *Value(stream_); }
+  const stream_t& Stream() const { return *Value(stream_); }
+  stream_t& operator*() { return *Value(stream_); }
+  const stream_t& operator*() const { return *Value(stream_); }
+  stream_t* operator->() { return stream_.operator->(); }
+  const stream_t* operator->() const { return stream_.operator->(); }
+
+ private:
+  void MasterFlipRequest(Request& r) {
+    if (r.method != "GET") {
+      r(current::net::DefaultMethodNotAllowedMessage(), HTTPResponseCode.MethodNotAllowed);
+      return;
+    }
+
+    if (!r.url.query.has("head") || !r.url.query.has("key")) {
+      r("", HTTPResponseCode.BadRequest);
+      return;
+    }
+    const bool has_i = r.url.query.has("i");
+    const auto client_head = std::chrono::microseconds(current::FromString<int64_t>(r.url.query["head"]));
+    const auto next_index = has_i ? current::FromString<uint64_t>(r.url.query["i"]) + 1 : 0;
+    const auto flip_key = current::FromString<uint64_t>(r.url.query["key"]);
+
+    const auto head_idxts = stream_->Data()->HeadAndLastPublishedIndexAndTimestamp();
+    if (client_head > head_idxts.head) {
+      r("", HTTPResponseCode.BadRequest);
+      return;
+    }
+    if (has_i && (!Exists(head_idxts.idxts) || next_index > Value(head_idxts.idxts).index + 1)) {
+      r("", HTTPResponseCode.BadRequest);
+      return;
+    }
+    if (Exists(head_idxts.idxts)) {
+      const auto next_index_by_timestamp =
```

Comment:
This condition with `std::min` reads black-magic-ish to me. A comment would help.